apis/changelogs/proto/v1alpha1/changelog.proto

```diff
@@ -0,0 +1,81 @@
+/*
+Copyright 2024 The Crossplane Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+syntax = "proto3";
+
+import "google/protobuf/struct.proto";
+
+// buf:lint:ignore PACKAGE_DIRECTORY_MATCH
+package changelogs.proto.v1alpha1;
+
+option go_package = "github.com/crossplane/crossplane-runtime/apis/changelogs/proto/v1alpha1";
+
+// ChangeLogService is a service that provides the ability to send change log
+// entries.
+service ChangeLogService {
+  // SendChangeLog sends a change log entry to the change log service.
+  rpc SendChangeLog (SendChangeLogRequest) returns (SendChangeLogResponse) {}
+}
```

Comment:
Interesting idea - I can see it making sense to minimize the number of calls/responses to the server in the change logs sidecar.  However, I'm having trouble understanding the lifecycle/management of the calls on the client side in the managed reconciler. Right now, the unary calls are directly tied to a single reconciliation - the call/response is complete within a single reconcile.

In a streaming approach, when does the client initiate the stream and then when would it decide to call `CloseAndRecv` to get the response from the server for the batch? 

The unary approach has a clear response within a single reconciliation, so we can easily attach it to the given resource before it finishes being reconciled. I'm not sure how to tie a response to a given reconcile in the streaming approach where we may not get a response until long after the reconcile is finished.

I'm sure we could think through how to manage the client side stream and getting responses reported appropriately for their corresponding resources, but do you feel like that extra complexity is worth it? Maybe it's more simple than I'm thinking 😇

Reply:
I've never used streaming gRPC before, but it my thinking is that you'd create the `stream` (i.e. the client with `Send` and `CloseAndRecv` methods) at provider startup. You'd then call `CloseAndRecv` only when the provider shuts down (so basically never).

You'd call the `Send` method on each reconcile to send a change log. This method returns an error, so presumably you'll get an error if the change log fails to send and can record that.

Given that AFAICT we don't really have a response for the server (sidecar) to send the client other than "yep I got your change log" I don't think it really matters if we never call `CloseAndRecv`.

This is all just based on a few minutes reading the gRPC streaming docs though, so 🤷‍♂️.

Reply:
I explored a client side streaming approach and I concluded that it's not a good fit for our scenario. Some artifacts of my exploration (please excuse the sloppy prototyping code 😇):

* https://github.com/jbw976/crossplane-runtime/commit/3754f69beee135226c74915e085bce84afbe81e3
* https://github.com/jbw976/provider-kubernetes/commit/29b4cc59fc9e3d6d5c12be6e57968ee72f7a7b83
* https://github.com/jbw976/change-log-sidecar/commit/59d09869bf5351a924d54e1388a10f31af0b9a76

Some observations from the prototype:

* I tried two different approaches for the client (reconciler) handing errors/messages from the server (change logs sidecar)
  * client only calls `Send()`: A server side internal processing error, as opposed to a transport/networking error, is not returned to the client from `stream.Send()`. A non transport error can only be received at the very end when the client finally calls `stream.CloseAndRecv()` and it only gets the first error the server encountered.
  * client calls `RecvMsg()` after `Send()`: The first time the client calls `RecvMsg()` it will get a full response and good server side internal error details back OK, but this closes the stream. Any subsequent attempt by the client to call `Send()` again will get `EOF`.
* bi-directional streaming could maybe work but that seems to be intended more for asynchronous sends/receives on different channels. IMHO we have a need for a classic call/response, where a single entry is sent and we get back an immediate single response. Just sticking with unary seems reasonable to me here 🤓

Reply:
Got it, thanks for investigating! Sounds like unary is the best approach here after all.